academy/3-my-own-chain/game-wager.md

```diff
@@ -405,17 +452,404 @@ With the desired steps defined in the wager handling functions, it is time to in
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/payment-winning/testutil/keeper/checkers.go#L21-L58]
+func CheckersKeeper(t testing.TB) (*keeper.Keeper, sdk.Context) {
+    return CheckersKeeperWithMocks(t, nil)
+}
+
+func CheckersKeeperWithMocks(t testing.TB, bank *mock_types.MockBankEscrowKeeper) (*keeper.Keeper, sdk.Context) {
+    storeKey := sdk.NewKVStoreKey(types.StoreKey)
+    memStoreKey := storetypes.NewMemoryStoreKey(types.MemStoreKey)
+
+    db := tmdb.NewMemDB()
+    stateStore := store.NewCommitMultiStore(db)
+    stateStore.MountStoreWithDB(storeKey, sdk.StoreTypeIAVL, db)
+    stateStore.MountStoreWithDB(memStoreKey, sdk.StoreTypeMemory, nil)
+    require.NoError(t, stateStore.LoadLatestVersion())
+
+    registry := codectypes.NewInterfaceRegistry()
+    cdc := codec.NewProtoCodec(registry)
+
+    paramsSubspace := typesparams.NewSubspace(cdc,
+        types.Amino,
+        storeKey,
+        memStoreKey,
+        "CheckersParams",
+    )
+    k := keeper.NewKeeper(
+        bank,
+        cdc,
+        storeKey,
+        memStoreKey,
+        paramsSubspace,
+    )
+
+    ctx := sdk.NewContext(stateStore, tmproto.Header{}, false, log.NewNopLogger())
+
+    // Initialize params
+    k.SetParams(ctx, types.DefaultParams())
+
+    return k, ctx
+}
+```
+
+The `CheckersKeeperWithMocks` function takes the mock in its arguments for more versatility.
+
+Now adjust the small functions that set up the keeper before each test. You do not need to change them for the _create_ tests because they never call the bank. You have to do it for _play_, _reject_, and _forfeit_.
+
+For _play_:
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/payment-winning/x/checkers/keeper/msg_server_play_move_test.go#L17-L32]
+func setupMsgServerWithOneGameForPlayMove(t testing.TB) (types.MsgServer, keeper.Keeper, context.Context,
+    *gomock.Controller, *mock_types.MockBankEscrowKeeper) {
+    ctrl := gomock.NewController(t)
+    bankMock := mock_types.NewMockBankEscrowKeeper(ctrl)
+    k, ctx := keepertest.CheckersKeeperWithMocks(t, bankMock)
+    checkers.InitGenesis(ctx, *k, *types.DefaultGenesis())
+    server := keeper.NewMsgServerImpl(*k)
+    context := sdk.WrapSDKContext(ctx)
+    server.CreateGame(context, &types.MsgCreateGame{
+        Creator: alice,
+        Black:   bob,
+        Red:     carol,
+        Wager:   45,
+    })
+    return server, *k, context, ctrl, bankMock
+}
+```
+
+This function creates the mock and returns two new objects:
+
+* The mock controller, so that the `.Finish()` method can be called within the test itself. This is the function that will verify the call expectations placed on the mocks.
+* The mocked bank escrow. This is the instance on which you place the call expectations.
+
+Both objects will be used from the tests proper.
+
+Do the same for [_reject_](https://github.com/cosmos/b9-checkers-academy-draft/blob/payment-winning/x/checkers/keeper/msg_server_reject_game_test.go#L17-L32).
```

Comment:
Should this also include a prompt to do the same for "forfeit" as well? Line 625 mentions "play, reject, and forfeit", but the guidance only mentions the first two.

Reply:
In forfeit it makes use of the "play" one. Adding this detail.